pkg/mutation/mutation_function.go

```
@@ -0,0 +1,220 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func Mutate(mutator Mutator, obj *unstructured.Unstructured) error {
+	return mutate(mutator, obj.Object, nil, 0)
+}
+
+func mutate(m Mutator, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch castPathEntry := pathEntry.(type) {
+	case *parser.Object:
+		currentAsObject, ok := current.(map[string]interface{})
+		if !ok { // Path entry type does not match current object
+			return fmt.Errorf("mismatch between path entry (type: object) and received object (type: %T). Path: %+v", current, castPathEntry)
+		}
+		next, ok := currentAsObject[castPathEntry.Reference]
+		if !ok { // Next element is missing and needs to be added
+			next = createMissingElement(m, currentAsObject, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case *parser.List:
+		elementFound := false
+		currentAsList, ok := current.([]interface{})
+		if !ok { // Path entry type does not match current object
+			return fmt.Errorf("mismatch between path entry (type: List) and received object (type: %T). Path: %+v", current, castPathEntry)
+		}
+		glob := castPathEntry.Glob
+		key := castPathEntry.KeyField
+		for _, listElement := range currentAsList {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else {
+				if listElementAsObject, ok := listElement.(map[string]interface{}); ok {
```

we should return an error if ok == false

---

actually, nvm, this works